### ISSUE TITLE
Restart browser if it became unresponsive

### ISSUE THREAD

**Issue by AndreyBelym** [DevExpress/testcafe#1815]
### Are you requesting a feature or reporting a bug?
Feature

### What is the current behavior?
When browser becomes unresponsive (`XXX browser disconnected. This problem may appear when a browser hangs or is closed, or due to network issues.`) we stop all running tests. 

### What is the expected behavior?
Unresponsive browser can be closed, and a new browser instance can be started on the next or even the same test (e.g. if quarantine mode is enabled).

### How would you reproduce the current behavior (if this is a bug)?
???

### Specify your

* testcafe version: 0.17.2
* node.js version: any

**Comment by pablorivera** [comment]
Hi, we are having the same issue with testcafe 0.17.2 with a couple of tests running in Bamboo CI that sometimes fails because of this error. It would be great if you can work on this ticket. Below is the stacktrace.

```
12-Oct-2017 10:43:40 | ERROR The Chrome 53.0.2785 / Linux 0.0.0 browser disconnected. This problem may appear when a browser hangs or is closed, or due to network issues.
-- | --
12-Oct-2017 10:43:40 |  
12-Oct-2017 10:43:40 | Type "testcafe -h" for help.
12-Oct-2017 10:43:40 |  
12-Oct-2017 10:43:40 | npm ERR! Linux 4.8.0-1.el7.elrepo.x86_64
12-Oct-2017 10:43:40 | npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "run" "test-ci" "tests"
12-Oct-2017 10:43:40 | npm ERR! node v6.7.0
12-Oct-2017 10:43:40 | npm ERR! npm  v3.10.3
12-Oct-2017 10:43:40 | npm ERR! code ELIFECYCLE
12-Oct-2017 10:43:40 | npm ERR! testcafe@1.0.0 test-ci: `testcafe 'chromium --no-sandbox' -s ./screenshots --screenshots-on-fails --color "tests"`
12-Oct-2017 10:43:40 | npm ERR! Exit status 1
12-Oct-2017 10:43:40 | npm ERR!
12-Oct-2017 10:43:40 | npm ERR! Failed at the testcafe@1.0.0 test-ci script 'testcafe 'chromium --no-sandbox' -s ./screenshots --screenshots-on-fails --color "tests"'.
12-Oct-2017 10:43:40 | npm ERR! Make sure you have the latest version of node.js and npm installed.
12-Oct-2017 10:43:40 | npm ERR! If you do, this is most likely a problem with the testcafe package,
12-Oct-2017 10:43:40 | npm ERR! not with npm itself.
12-Oct-2017 10:43:40 | npm ERR! Tell the author that this fails on your system:
12-Oct-2017 10:43:40 | npm ERR!     testcafe 'chromium --no-sandbox' -s ./screenshots --screenshots-on-fails --color "tests"
12-Oct-2017 10:43:40 | npm ERR! You can get information on how to open an issue for this project with:
12-Oct-2017 10:43:40 | npm ERR!     npm bugs testcafe
12-Oct-2017 10:43:40 | npm ERR! Or if that isn't available, you can get their info via:
12-Oct-2017 10:43:40 | npm ERR!     npm owner ls testcafe
12-Oct-2017 10:43:40 | npm ERR! There is likely additional logging output above.
12-Oct-2017 10:43:40 |  
12-Oct-2017 10:43:40 | npm ERR! Please include the following file with any support request:
12-Oct-2017 10:43:40 | npm ERR!     /apps/testcafe/npm-debug.log
12-Oct-2017 10:43:41 | Tests failed
```

**Comment by pablorivera** [comment]
It also happens with testcafe 0.18.1

**Comment by gtroshinnc** [comment]
Annoying issue. Any ETA here? :) 0.18.5 still has it

**Comment by berrutti** [comment]
I'm having the same problem. I thought it was related to local execution but our client experiences the same on his machine.

**Comment by AlexanderMoskovkin** [comment]
Hi,

Thanks for your interest. This issue is in our plans to the one of the next releases.

**Comment by guptaakhil** [comment]
My observation is here if we run this in bulk run then only face this issue. @AlexanderMoskovkin  : what is your thought on this.

**Comment by sijosyn** [comment]
Can we get this feature implemented in ASAP - @AndreyBelym / @AlexanderMoskovkin? This issue is very annoying and blocking our test execution on Safari, Firefox (& maybe Edge/IE). We are still seeing this issue in the latest TestCafe version 0.20.0-alpha.1 ? Please let me know if I can help provide any additional debug logs. I already provided some logs here  - https://testcafe-discuss.devexpress.com/t/browser-disconnected-error-this-problem-may-appear-when-a-browser-hangs-or-is-closed-or-due-to-network-issues/383

**Comment by kirovboris** [comment]
Hi @sijosyn, we plan to implement this feature in next sprint (0.21.0 version). We'll let you know when the dev build will available.

**Comment by guptaakhil** [comment]
I have face same issue with browsers stack site when I am executing my
application over there.

**Comment by guptaakhil** [comment]
Any ETA for it's fix?

**Comment by dmc1522** [comment]
Any update on this? I have a couple of suites running on Jenkins and randomly stop because of this... I am really thinking about creating a daemon to keep checking the job and re-start everything in case it hangs like this, but that doesn't seem like a good solution...  is this going out already?

**Comment by berrutti** [comment]
Is there a way to contribute code to this issue? I would like to help in any way possible. Thanks

**Comment by AndreyBelym** [comment]
Hi @berrutti, any contributions are welcome. I think all things that should be changed lies on [https://github.com/DevExpress/testcafe/blob/master/src/browser/connection/index.js](src/browser/connection/index.js), look at the [_waitForHeartbeat](https://github.com/DevExpress/testcafe/blob/master/src/browser/connection/index.js#L121) function. If you make a PR, I can help with tests and can answer any questions that will arise.

**Comment by miherlosev** [comment]
https://github.com/DevExpress/testcafe/issues/1815

TestCafe should restart the browser when the latter starts to slow down.

I suggest using the following algorithm:
* implement a user-defined option to specify the memory usage limit
* track the browser memory usage (only for IE) and restart the browser when it exceeds this limit

**Comment by SergeyZhurov1992** [comment]
Hello!
I'm also very need this feature: 
track the browser memory usage (only for IE) and restart the browser when it exceeds this limit

**Comment by crhistianramirez** [comment]
What kind of workarounds have people implemented? We saw this on IE only and so we are thinking of running firefox and chrome in parallel and then IE after that in series but if the issue is affecting all browsers as some are reporting then I dont see it working.

**Comment by Sashikus** [comment]
Hello, we have the same issue for Chrome and Firefox on CircleCI with Testcafe v.1.0.1, any updates on this issue? thanks

**Comment by valerii-cognite** [comment]
Same issue here on `chromium:headless` with ` --no-sandbox` flag. Happening occasionally in different tests.

**Comment by AndreyBelym** [comment]
@valerii-cognite, have you tried the `--disable-dev-shm-usage` Chrome flag?

**Comment by valerii-cognite** [comment]
@AndreyBelym thank you for the suggestion. Since the problem reproduces occasionally I tried to run my tests 20 times in a row with the `--disable-dev-shm-usage` flag set. It appeared that two of those ended up with the same problem of a browser being disconnected. So I would conclude that the flag doesn't help.

**Comment by AndreyBelym** [comment]
@valerii-cognite have you considered creating a [new bug report](https://github.com/DevExpress/testcafe/issues/new?template=bug-report.md) and providing an example?

**Comment by valerii-cognite** [comment]
@AndreyBelym unfortunately I can't reproduce the problem locally but I created an issue where I submitted all the known information so far https://github.com/DevExpress/testcafe/issues/3650

**Comment by itdoginfo** [comment]
I want to add another example why this function is needed.

The example contains testing 3D graphics on mobile devices. Not all mobile devices can load complex 3d graphics.

This is a simplified test to make it easy to reproduce:
```
npm install testcafe-browser-provider-android
```

test.js
```
import { Selector } from 'testcafe';

var examples = ['https://devexpress.github.io/testcafe/example/', 'https://xeokit.github.io/xeokit-sdk/examples/#sceneRepresentation_PerformanceModel_batching_benchmark', 'google.com'];

examples.forEach(example => {
   
    fixture `Test page: ${example}`
        .page (`${example}`);

    test('My test', async t => {
        const element     = Selector('#developer-name', {timeout:90000});
        const clientWidth = await element.clientWidth;
    });
});
```
A timeout is needed to wait until browser falls on the 3d.

```
npx testcafe android::52109133:chrome test.js
```

The browser cannot load the 3D model and the testcafe restarts this test after some time, and so on three times. After that, the testcafe completes its work with output:
```
ERROR The Chrome Mobile 76.0.3809 / Android 8.0.0 browser disconnected. This problem may appear when a browser hangs or is closed, or due to network issues.
```

In the example, several tests and I expected that after a failed test on one page, the tests will continue on the next. But no, this completely interrupts the work.

I use Samsung Galaxy A5 (2017) and webdriverio + appium as browser-provider. And I get the same issue on ios devices.

**Comment by AlexKamaev** [comment]
https://github.com/DevExpress/testcafe/issues/1815 


@itdoginfo
The purpose of this feature is improving the stability of tests under the unexpected conditions which make the browser hang.
This means that if the browser hung during the test, TestCafe tries to restart the browser on this specific test, because TestCafe does not exactly know that your heavy page leads to hanging.  It's possible that your machine just does not have enough memory because of some other application. In this case, we try to restart the browser to finish the test.
In other words, this feature must help the test to pass and not to fail it. If you know exactly that some of your tests is very heavy for some devices, it's better to exclude this test from executing on these devices, and not relying on the mechanism which is not intended for this.

**Comment by itdoginfo** [comment]
@AlexKamaev 
I almost agree with current behavior, the only difference I expect testcafe to fail test and move on to the next one.

>In other words, this feature must help the test to pass and not to fail it. If you know exactly that some of your tests is very heavy for some devices, it's better to exclude this test from executing on these devices, and not relying on the mechanism which is not intended for this.

We have test suite with a kind of stress tests, that we try to run on as many devices as we can.
We are expecting tests to be as heavy as they can.
Current behavior is ok, we see that not all of our tests passed, but if we failed on the first test, we do not know the state of other tests.

Maybe there is some workaround?

**Comment by AndreyBelym** [comment]
TestCafe was not designed with stress testing support in mind, so I'm afraid it's not possible to achieve your goal without heavily modifying the TestCafe code. But I think we can make stress testing possible once we allow using TestCafe for browser automation tasks - https://github.com/DevExpress/testcafe/issues/2501.

**Comment by kirillgroshkov** [comment]
We're not using TestCafe for stress-testing, but still our e2e tests hang with "browser disconnected" message every now and then, which makes them very unstable. We're almost considering dropping TestCafe because of this single issue, cause it was very annoying throughout last year. We're managing some sophisticated "restarting scripts" right now to mitigate that, but that increases our e2e run time by 2 / 3 times (cause restart restarts ALL tests).

**Comment by rehael** [comment]
How hard would it be to add something along the lines of `test.skipOnDisconnect()` (or any other method of tagging the test with specific instructions for the runner how to handle the test – maybe in _meta_ like `test.meta({skipOnDisconnect: true})`?), which upon browser disconnect would normally restart browser, and continue from the _next_ test? Where should I start looking if I'd like to create a PR for it?

**Comment by kirillgroshkov** [comment]
Or `test.meta({restartOnDisconnect: true})`, since for us it eventually works after 1/few restarts, and would be more efficient than restarting ALL tests

**Comment by AndreyBelym** [comment]
@kirillgroshkov such behavior clearly indicates that TestCafe conflicts with your pages in some way. Even if we skip a test that caused a browser failure, there is no guarantee that the next test will pass. It's better to create a separate bug report and provide an example or access to your pages. In this case, we can effectively debug and fix the reason that causes browser disconnections.

@rehael a change like this will cover several subsystems (compiler, browser connections, testing runtime) and due to its complexity, we will need to discuss if we really want to include such patch in our codebase.

**Comment by AndreyBelym** [comment]
I close this issue since restarting browsers that don't respond was implement in a basic way. Enhancements suggested by @miherlosev () are extracted to a separate issue: https://github.com/DevExpress/testcafe/issues/4132. The suggestion about skipping the test that caused the DNR state in the browser is extracted to https://github.com/DevExpress/testcafe/issues/4133. If you have any other suggestions about browser restarting, please don't hesitate to create [new feature requests](https://github.com/DevExpress/testcafe/issues/new?template=feature_request.md).
